xformers/factory/model_factory.py

```diff
@@ -155,21 +155,22 @@ def __init__(
                     self.reversible_encoder = config.reversible
 
             # Build up the stack
-            for i in range(config.num_layers):
+            num_layers = config.num_layers + 1
+            for i in range(num_layers):
                 # Label where this layer is in the stack
                 # (for instance useful for the positional encoding, or late layer norm)
                 if len(recipient) > 0:
                     config.layer_position.mark_not_first()
 
-                if config != stack_configs[-1] or i < config.num_layers - 1:
+                if config != stack_configs[-1] or i < num_layers - 1:
                     config.layer_position.mark_not_last()
 
                 # If reversible: extract the reversible sub-parts, else append the block as-is
                 if self.reversible_encoder and i > 0:
                     f, g = xFormerEncoderBlock.get_reversible_layer(config)
                     recipient.append(torch.nn.ModuleList([f, g]))
                 else:
-                    recipient.append(builder(config))  # type: ignore
+                    recipient.append(builder(config, compute_attention=False))  # type: ignore
```

Comment:
This screws the non-reversible arch, doesn't it? IAnyway,  would do it differently:

```python
if self.reversible_encoder:
    if i == 0:
        recipient.append(torch.nn.Sequential([
            ...embedding from the config...,
            rv.InputAdapter(),
        ])
    f, g = xFormerEncoderBlock.get_reversible_layer(config)
    recipient.append(torch.nn.ModuleList([f, g]))
else:
    recipient.append(builder(config))
...
grouped_encoders = [encoders[0]]  # changed
for i in range(1, len(encoders) + 1, self.reversible_encoder):  # changed
```

This is way less intrusive than adding a sneaky flag and rewriting the block factory.

Reply:
Why +1 on len(encoders)?

Reply:
The first element is the InputAdapter. It's required to convert from `[batch, sequence, embedding]` to `[2 * batch, sequence, embedding]` of the reversible chain.

Reply:
The first element in `encoders` is the new `xFormerEmbeddingBlock` and the second is `InputAdapter`, and the rest are reversible layers (f and g). Here from what I see is about grouping every `self.reversible_encoder` revisble layers into one `ReversibleSequence`. Therefore the number of revisible layers to be grouped should be `len(encoders) - 2`. So, the iteration should start from index 2, till the end of `encoders`.

```
index:    [xFormerEmbeddingBlock InputAdapter (f,g) (f,g) (f,g) ...]
encoders: [         0                 1         2     3     4   ...]
```

The code should be:
```py
grouped_encoders = [encoders[0], encoders[1]]
for i in range(2, len(encoders), self.reversible_encoder):
    ...
```

`i` should never be greater than `len(encoders)`, because it is the index of the starting reversible layer of the new `ReversibleSequence` in `encoders` list.